### ISSUE TITLE
passing invalid long option to `function` silently causes function definition to fail

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#2827]
While responding to this [comment](https://github.com/fish-shell/fish-shell/issues/2036#issuecomment-198518520) I noticed that passing an invalid long option to the `function` command causes the function definition to silently fail.
### Reproduction Steps:
1. Type

```
function wtf --on-function-return
    echo wtf called
end
```

The `$status` var is set to one but no error message is emitted.
1. type `wtf` and notice you get this error: `fish: Unknown command 'wtf'`
2. type `functions` and notice **wtf** isn't present.
### Expected behavior:

The function definition fails with an explicit error message.

---

Fish version: git head

**Comment by marmistrz** [comment]
This is not present in 2.2.0. I typed this myself and I get the `function` manual in return. So probably a regression.

**Comment by cjappl** [comment]
Seconded @marmistrz, I just get the manual as well

**Comment by krader1961** [comment]
Removed the easy-pick label because in looking at the code fixing this will not be easy.

**Comment by krader1961** [comment]
_I posted the following comments to Gitter but decided they should be in this issue as well to ensure that they are noticed by more people._

The function man page makes no mention of allowing the `-a`/`--argument-names` flag before the function name. Yet the implementation goes out of its way to accommodate that form. Why? I'd really like to simplify the `builtin_function()` arg processing so it uses the same approach to arg parsing as the other builtins.

Ah! I forgot that the function -a flag has weird semantics. But the implementation is still wrong IMO because it allows things like `function testing -d "testing" def -a ghi; end` where both `def` and `ghi` get treated as named args. Worse, if you do `function -d 'hello' -a def ghi wtf; end` then the function name is taken as the last positional argument (`wtf` in this example). This logic could be greatly simplified if we always required the function name to be the first arg and `-a` has to precede the first non-flag arg.

Doing it that way also makes it easy to support the more natural syntax for specifying named args; e.g., `function testing -a var1 -a var2`. We would still, obviously, need to allow any unbound args to be treated as named args if `-a` has already been seen. But that logic would be much simpler than it is now.

**Comment by krader1961** [comment]
There should not be any instances of people doing `function -a arg1 arg2 func_name` so I'm going to rewrite the code to disallow that. The function name has to be the first argument. After that the usual option parsing rules are in effect. Plus, for backward compatibility, any unbound non-flag args will be associated with the `-a` flag if it was already seen; otherwise, they will cause an error since they aren't associated with a flag.

**Comment by ridiculousfish** [comment]
I like your proposed simplification. #2068 relevant. See 286c60bc9d1b692b650bd0c37a6da635e99f119f if you want to undo what I did; you may be able to just revert that commit.

**Comment by krader1961** [comment]
Yes, PR #2068 is relevant as it seems to be the trigger for this regression. Sadly the author of that PR did not open an issue where the problem could be discussed. Which lead you to not just merge the change to the documentation but to also modify the implementation to explicitly support what was implied by the previous documentation. We really need to stop accepting non-trivial pull-requests that are not tied to an issue.

**Comment by krader1961** [comment]
Also, the change I'm about to merge will break the explicit `function` signature prior to PR #2068 and the implicit signature since then. Which is a backward incompatible change. I think the risk is warranted. It is very unlikely that anyone is defining functions by putting flags before the function name and if they are doing so they, frankly, deserve to be broken for taking advantage of that implementation bogosity.

**Comment by krader1961** [comment]
For the record the change I just merged alters the semantics of `-a`. It used to be that this

```
function myfunc abc -a def
```

was equivalent to this

```
function myfunc -a abc def
```

With my change it is equivalent to `function myfunc -a def abc` (note the inversion). That's because I modified the flag parsing to make `-a` require an argument. We shouldn't allow the first form but that would greatly complicate the logic. So instead I'm going to merge another change that makes `-a` not bind to the next argument and instead just use all unprocessed positional args in their original order as the bound argument names. That will restore the original, ugly, semantics. I think it highly unlikely anyone is depending on it but there's no reason to make that particular backward incompatible change. Obviously the function name must still be the first argument -- which is strictly speaking not backward compatible but as I said earlier anyone depending on that deserves to be broken.
